[QoI] Handle out of order arguments by reordering them

d-ronnqvist opened the pull request:
Change handling of out of order arguments to reorder all arguments according to the parameter order instead of just performing one swap.

I'm not sure if this change is in the spirit of the original `outOfOrderArgument` method or not and if it's what @dabrahams had in mind for SR-4715 or not. 

As I see it, the new diagnostic message is more general - for better and worse - but the fix-it can rearrange multiple arguments at once. In many of the existing test cases where there are only two arguments that need to be swapped this new can be a step back in terms of clarity. At the same time I find that seeing listing all the arguments can provide additional context.

As for the fix-it; in simple cases it's likely unnecessary to modify all of the arguments, but it allows this changed code to deal with many arguments that are completely out of order.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-4715](https://bugs.swift.org/browse/SR-4715)

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

d-ronnqvist committed:
[QoI] Improve the fix-it for out-of-order arguments.

Change the fix-it to move the argument to its correct location in one go. This happens by removing it from one location and inserting it in the other (as opposed to the original implementation which swapped one argument with the preceding one). The commas separating the arguments are adjusted to match the moved argument.

Add new tests for reordering regular arguments, variadic arguments, and function arguments.

Resolves: SR-4715 rdar://problem/31849281

xedin reviewed (changes requested):
Thank you for the patch, @d-ronnqvist!

I think it would be great if we could use the old style errors + fix-its but do it right instead of asking users to drag label through every argument position, which is what is happening right now unfortunately :( It seems to me that errors, in their current form, are more "actionable" and remove some of cognitive load from users comparing to the new error messages, where users would have to parse label string to figure out differences between right and wrong positions, having fix-it which replaces all of the arguments together doesn't help either. 

Could you please add couple more test-cases were arguments are more complex than single integer or string literals e.g. function calls and/or closures?

d-ronnqvist commented:
@xedin How about this? Now it has the same error message as it previously did, but the fix-it moves the out-of-place argument into the correct location in one go. The only "extra" source manipulation is to move a ", " along with the moved argument.

I also added a few more new tests but wasn't entirely sure what to test. Suggestions are appreciated. (`test/Constraints/keyword_arguments.swift` also already covers some scenarios with default arguments and variadics.)

I rebased my changes. Sorry if the review requests got lost in doing so.

xedin commented:
@d-ronnqvist Awesome! I'll take a closer look soon, just need to get few things done first. Thanks, again!

xedin commented:
@swift-ci please smoke test

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -449,50 +449,45 @@ matchCallArguments(ArrayRef<AnyFunctionType::Param> args,
-        continue;
-      }
+    unsigned argIdx = 0;
+    // Enumerate the parameters and their bindings to see if any arguments is our of order
```
typo -> is/are

xedin commented on test/Constraints/diagnostics.swift:
```diff
@@ -1060,3 +1060,26 @@ class ListExpr_28456467 : AST_28456467, Expr_28456467 {
+// expected-error @-1 {{argument 'b' must precede argument 'c'}} {{26-26=b: { (num: Int) -> String in return String(describing: num) }, }} {{38-101=}}
+fun_31849281(a: { !$0 }, c: [nil, 42], b: { "\($0)" })
+// expected-error @-1 {{argument 'b' must precede argument 'c'}} {{26-26=b: { "\\($0)" }, }} {{38-54=}}
+
```
This is the example I've made which had a broken fix-it before but looks correct now:

```
func foo(x: Int, y: Int) {
}
foo(42, x: 0)
```

xedin replied:
Actually I'm not sure if its correct because it highlights the whole region, can you please verify?

d-ronnqvist replied:
I think it's correct in highlighting the range that's being removed and the unlabeled argument it needs to precede. Having one argument in between shows that it's not the whole region:
```
(swift) func bar(x: Int, y: Int, z: Int) {}
(swift) bar(42, y: 10, z: 20, x: 0)
error: argument 'x' must precede unnamed argument #1
bar(42, y: 10, x: 0)
    ~~       ~~^~~~
    x: 0, 
```
But when there's only two arguments the two ranges meet and it looks like the whole range. It might look better if it didn't highlight the preceding comma in the range that's being removed. But I don't know if the highlight when using `fixItRemove` can be changed.

xedin replied:
Sounds good, please add it to the tests for completeness.

xedin reviewed (approved):
Overall LGTM now, please fix the typo I've mentioned, squash the fix-it commit and format the commit description to cut-off lines at 80 symbols or so and we should be good to go.

xedin commented:
@d-ronnqvist And I forgot to mention - could you also please run `clang-format` on the changes you've made, we have a tool which integrates it with git too, so you could do `git clang-format`. It's located in `clang/tools/clang-format`, binary itself should be in the `bin` directory of your `llvm` build which comes with swift.

d-ronnqvist commented:
@xedin Did you want me to squash all commits or only the modifications to the second commit (this solution)?

xedin commented:
@d-ronnqvist Sorry, please squash all 3 into a single one, I'm going to nominate it for 4.0 as well.

xedin commented:
@swift-ci please smoke test

xedin commented:
@d-ronnqvist Thank you!
